[OLD] Replace XQuartz-based GUI forwarding with headless X11 and noVNC

matejstastny opened the pull request:
## Summary

This PR makes a new and working solution on how to run Gazebo on MacOS. Previous attempts using `XQuartz`-based `X11` forwarding did not work at all. This is the first approach that reliably launches and displays Gazebo.

The new setup runs a complete headless desktop environment inside the container and exposes it through `noVNC`, making the GUI accessible directly from a web browser.


## Headless X11 + VNC + noVNC

This PR replaces `XQuartz` entirely with a container-native GUI stack that runs fully inside Docker:

Inside the container:

- `Xorg`: creates a virtual X11 display
- `Openbox`: provides a lightweight window manager
- `x11vnc`: exposes the display via VNC
- `noVNC`: serves a browser-based VNC client

All rendering happens inside the container, eliminating the need for any host-side X11 server or display forwarding.

This is the first configuration that works end-to-end for running Gazebo in the devcontainer on MacOS.

The new workflow is also very user-friendly:

- Requires only a web browser
- No `XQuartz`
- No `X11` setup
- No VNC client installation

### Usage:

1.	Start the `X11` stack inside the container:

```bash
./ros2_ws/scripts/start_x_server.sh
```
	
2.	Launch Gazebo:

```bash
ign gazebo empty.sdf
```

4.	Open [http://localhost:6080/vnc.html](http://localhost:6080/vnc.html) and click connect

## Other changes

- I documented the `Dockerfile` and the `start_x_server.sh` script so everybody knows what both of them are doing
- I repurposed the `run_macos.sh` script from being an XQuartz launcher and X11 configurator to be a general attach script for all platforms for when you don't want to use the VsCode terminal but other external terminal

RadAlpaca11 committed:
updating windows instructions

matejstastny committed:
Added a script to export docker image

matejstastny committed:
Merge branch 'macos-setup-matej' of https://github.com/TrickfireRobotics/gazebo-simulations into macos-setup-matej

matejstastny committed:
Added display var log

matejstastny committed:
Formatted instruction comments

matejstastny committed:
Removed display env var from devcontainer

matejstastny committed:
Added new display method stuff to Dockerfile

matejstastny committed:
Removed old unused script

matejstastny committed:
Repurposed run script into a plain container attach script

matejstastny committed:
Fixed script description

matejstastny committed:
Added xorg config

matejstastny committed:
FINALLY Removed libGl env var making gazebo work ( it was a problem in my script all along )

matejstastny committed:
Properly documented the Dockerfile, removed x11 tools, because they are no longer needed

matejstastny committed:
Modified attach path of the attach script to match vscode default attach path

matejstastny committed:
Added proper documentation to the x server script

matejstastny committed:
Added noVnc to connect from browser

matejstastny committed:
Updated README run instructions

matejstastny committed:
Merge branch 'main' into macos-setup-matej

matejstastny committed:
Added summary and other logs

matejstastny committed:
Merge branch 'macos-setup-matej' of https://github.com/TrickfireRobotics/gazebo-simulations into macos-setup-matej

matejstastny committed:
Removed broken summary

matejstastny committed:
Removed hardcoded ports from x server script

matejstastny committed:
Fixed incorrect script name in attach script header

knguyen-hub commented:
Hey I just reviewed this, it seems to work for me on Windows. I'm going to take a closer look at the code changes before approving

matejstastny committed:
Removed useless suggestions from variable checks

matejstastny committed:
Removed output from commands with the option of enabeling it using the verbose flag

matejstastny committed:
Made it so that xorg openbox x11vnc and novnc gets killed on script exit to prevent background processes

matejstastny commented:
## NEW CHANGES

I added some quality of life features, because I didn't like some stuff while working with it.

### Command output

I added a cyan to the logs the script makes, because I was getting lost in the output of the commands, and couldn't find my logs.

I made it so the command output is not shown, because the `x server` and `vnc server` have an extremely verbose output. It is possible to enable the output by using the `-v` or `-verbose` flag. The command output will get shown any of the commands fail for debugging.

The log is also saved in `/tmp/start_x_server.log` if you wanted it as a file.

### Process termination

It was really hard to test the script, because every time I ran it the processes started in background even after I ended the script. The script now traps a cleanup that will kill all processes created by the script, and also it will wait until the script is terminated by the user.

matejstastny committed:
Fixed hardcoded port in novnc log message

matejstastny committed:
Moved all scripts into one script directory

matejstastny committed:
Moved xorg config into the devcontainer dir because its a container file like the bashrc

matejstastny committed:
Renamed attach script to have a more descriptive name

knguyen-hub commented on .devcontainer/Dockerfile:
```diff
@@ -1,106 +1,203 @@
+EXPOSE 6080 5900
+
+# Environment variables used by X11, VNC, and helper scripts
+ENV DISPLAY=:1
```
```suggestion
ENV DISPLAY=:0

anonymousomeone commented:
lgtm

matejstastny committed:
Changed targeted display from 1 to 0

Co-authored-by: knguyen-hub <54388203+knguyen-hub@users.noreply.github.com>

matejstastny committed:
Merge pull request #6 from TrickfireRobotics/fix-project-structure

Adressed bad file structure
